### PR TITLE
modify server table to allow duplicate entries, increase char length

### PR DESCRIPTION
**Pull request by dg4prez**
This will allow duplicate IP and hostname entries, as well as increase the character limit from 45 to 63, which is the max allowed per segment according to rfc.  This also removes the DB requirement to have subnet mask, port name, and gateway.  The UI still requires this - items like ORG should be modified to not require these entries in the UI.

### CODE REVIEW

**Comment by dg4prez** [comment]
https://github.com/Comcast/traffic_control/issues/1619

**Comment by mitchell852** [comment]
if you change your previous comment to 'fixes #issueNumber', the issue will be closed when the PR is merged...if that's what you want...

**Comment by mitchell852** [comment]
Removing constraints from the TO database (i.e. a unique constraint or a not null constraint) could potentially harm the integrity of the data upon which the CDN relies. To me, it seems like without the unique constraint, you could create a cachegroup with servers in it like this:

server A: 2.2.2.2
server B: 2.2.2.2
server C: 2.2.2.2

which may not break the CDN per se but will cause it to perform at a suboptimal level (load balancing is basically turned off). Maybe there are other implications that I am not aware of. I just want to make sure there are no unwanted side effects of this change. Anyone else have any thoughts or concerns?

**Comment by mitchell852** [comment]
so a couple of things off the top of my head needs to be done in addition to the database schema changes:

1. it sounds like UI modification is needed to not require subnet mask, port name, and gateway if server.type == ORG on server create/update
2. API modification is needed to not require subnet mask, port name, and gateway if server.type == ORG on server create/update
3. test cases
a. testcase to test that duplicate IP and hostname entries are accepted and persisted to the db
b. testcast to test # 2 (create or update an ORG server thru the api without subnet mask, port name, or gateway)

**Comment by dg4prez** [comment]
Are there any issues with keeping the duplicate hostnames change?  We need either that or to merge the hostname into a single FQDN.  We have encountered cases where origins have identical hostnames but different domains, especially in cases where we are setting up for dual-origin configs.

**Comment by mitchell852** [comment]
It makes sense to me that server.host_name should not be unique as you may have 2 origin servers that look like:

foo-bar.domain1.com
foo-bar.domain2.com

how about instead of dropping the unique constraints altogether, add multiple fields to the constraints like this maybe (i think this would work but i'm no DB expert):

```alter table server drop index host_name;

ALTER TABLE `server` ADD UNIQUE `unique_index`(`host_name`, `domain_name`, `cdn_id`);

alter table server drop index cs_ip_address_UNIQUE;

ALTER TABLE `server` ADD UNIQUE `unique_index`(`ip_address`, `cdn_id`);

alter table server drop index ip6_address;

ALTER TABLE `server` ADD UNIQUE `unique_index`(`ip6_address`, `cdn_id`);```

**Comment by mitchell852** [comment]
also, you'll want to check that these indexes actually exist before attempting to drop them. dropping an non-existent index will cause the migration to fail.

**Comment by mitchell852** [comment]
i made some changes to the migration file, however, the UI and presumedly the API still enforces interface_name, ip_netmask and ip_gateway as required fields. i'll hold off on merging this until that is done. i can chip in on that next week.

File: docs/source/development/traffic_ops_api/v12/server.rst [hunk 1]
```diff
@@ -661,13 +661,13 @@ Server
   +------------------+----------+------------------------------------------------+
   | cachegroup       | yes      | cache group name                               |
   +------------------+----------+------------------------------------------------+
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/server.rst]: change these back to required

File: traffic_ops/app/db/migrations/20160826111234_modify_server_table.sql [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+PREPARE stmt2 FROM @sqlstmt2;
+EXECUTE stmt2;
+
+set @exist3 := (select count(*) from information_schema.statistics where table_name = 'server' and index_name = 'host_name' and table_schema = database());
+set @sqlstmt3 := if( @exist3 > 0, 'alter table server drop index host_name', 'alter table server');
+PREPARE stmt3 FROM @sqlstmt3;
+EXECUTE stmt3;
+
+alter table server modify host_name varchar(63) not null;
+alter table server modify domain_name varchar(63) not null;
```
Comment (mitchell852) [inline traffic_ops/app/db/migrations/20160826111234_modify_server_table.sql]: change interface_name, ip_netmask and ip_netmask back to not null

File: traffic_ops/app/lib/API/Server.pm [hunk 3]
```diff
@@ -412,18 +413,9 @@ sub check_server_params {
 	if ( !defined( $json->{'cachegroup'} ) ) {
 		$errFields{'cachegroup'} = 'is required';
 	}
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Server.pm]: put these lines back in

**Pull request merged**